### PR TITLE
fix: prevent negated patterns from traversing directory boundaries

### LINKED ISSUE #40: Traverses directory boundaries when performing negative matching
Hi!

I have noticed surprising behaviour when running a set of integration tests that I put together against `micromatch` (a javascript glob algorithm). I am on board with almost all of the changes, _except one_, and I wanted to make sure it was a bug before diving into the code to 'fix' it.

When presented with the glob `a[!b]c`, `wax` will happily match path separators in place of `b` which is odd. So, `a/c` passes that test case.

```rust
    #[test]
    fn negative_match_does_not_traverse_folders() {
        let glob = Glob::new("a[!b]c").unwrap();
        assert!(glob.is_match(Path::new("adc")));
        assert!(!glob.is_match(Path::new("a/c")));
    }
 ```
 
 Compare this to a glob playground:

https://www.digitalocean.com/community/tools/glob?comments=true&glob=a%5B%21b%5Dc&matches=false&tests=%2F%2F%20expected&tests=adc&tests=%2F%2F%20wax%20fails%20here&tests=a%2Fc

### PR DESCRIPTION
**Pull request by arlyon**
Closes #40 by not pushing &&[^/] if we are already in a negation pattern, and just pushing the separator instead.

### CODE REVIEW

File: src/encode.rs [hunk 1]
```diff
@@ -250,7 +250,13 @@ fn encode<'t, A, T>(
                             },
                         }
                     }
-                    pattern.push_str(nsepexpr!("&&{0}]"));
+                    if class.is_negated() {
```
Comment (olson-sean-k) [inline src/encode.rs:253]: This code now repeats the same branch predicate (`class.is_negated()`) with some common code in between. I'd prefer to factor the common code into a function and use only one branch:

```rust
...
use crate::token::Class as ClassToken;

fn encode_class_archetypes(class: &ClassToken, pattern: &mut String) {
    for archetype in class.archetypes() {
        ...
    }
}

let mut pattern = String::new();
pattern.push('[');
if class.is_negated() {
    ...
    encode_class_archetypes(&class, &mut pattern);
    ...
}
else {
    ...
    encode_class_archetypes(&class, &mut pattern);
    ...
}
pattern.push_str("]");
...
```

File: src/lib.rs
Comment (olson-sean-k) [inline src/lib.rs:1]: Thanks for including these tests! :smile: I have a few nitpicks that all concern formatting in some way. I've tried to order them from what I think are less nitpicky to probably very nitpicky.

- I think these tests should be placed nearby existing tests concerning character classes (e.g., `match_glob_with_class_tokens`). At time of writing on `master`, these tests begin near `lib.rs:1650` or so. Somewhere at the end of the existing tests would work well in my opinion.
- The names of tests currently lack an expectation, which isn't great. Despite that, I'd prefer the names of these tests to be consistent with other tests as written today, so something more like `match_glob_with_only_negated_character_class_tokens` and `match_glob_with_only_negated_range_class_tokens`. I have some plans to improve this and some other issues with tests that I hope to put together later.
- Other tests use blank lines to separate cases, test data, positive assertions, and negative assertions. Following that pattern, these tests would have blank lines between each of their three lines of code. I realize all that extra space is probably controversial here... 😛

**Review by olson-sean-k — CHANGES_REQUESTED** [review]
Thanks for the fix! This looks good aside from one refactor that I'd prefer and some style nitpicks regarding the tests.

**Comment by olson-sean-k** [comment]
Thanks for the PR and sorry for the radio silence! If you're still willing to land this, please let me know and take a look at the review. I realize this change was requested some time ago and I completely understand if you can't or don't want to get this landed now.

File: src/encode.rs (hunk 1 above)
Comment (olson-sean-k) [reply src/encode.rs]: I've refactored this in an additional commit in this PR.

File: src/lib.rs
Comment (olson-sean-k) [reply src/lib.rs:1]: I've made this change in an additional commit in this PR.

**Comment by arlyon** [comment]
Hey! Apologies for just abandoning this, been very busy recently. I appreciate you taking the time to merge this :tada:

**Pull request merged**